### DISCUSSION TITLE
Ship Viewscreen

### DISCUSSION THREAD

**Discussion by alexanderson1993** [Thorium-Sim/thorium-nova#320] [Ideas]
Once we get some engine and thruster ship systems and some controls in place for adjusting those systems, we can start building the viewscreen. This discussion is about all the stuff that shows up on the view screen. I'm going to start this discussion with some first-principles, but I mostly want to hear what everyone else thinks.

### The Viewscreen's Purpose

The view screen serves as a way for the crew to get context about what's happening in the broader mission. From my experience as a CMSC flight director, during the flight it's possible for crew members to get a little stuck doing their individual job - which is a good thing! We want each station to feel meaningful and engaging. But the jobs are in service of the story, and if crew members are missing the story, they're missing a lot.

The viewscreen is a landmark that brings the crew back to what is currently happening. They should be able to watch what's happening on the viewscreen for a few moments to get an idea of what is going on right then.

### How Thorium Classic space centers use viewscreens

When I was at the CMSC, the viewscreen primarily showed one of three things:

- Generic views for when there was nothing else to show, like random computer screens when the ship was still, warp stars when the ship was traveling, various generic graphics and animations about the current status of the ship, etc.
- A tactical display, which was like a glorified powerpoint presentation, with a separate presentation for each mission. It would show animations and graphics depicting what was happening around the crew.
- Contextual video related to the events of the mission. This was a DVD clipped together from various TV shows and movies. It included ship moving and battling in space, explosions rocking the inside of the ship when the crew's ship was hit, intruders moving around inside the crew's ship, etc. 

It took a lot of time and effort to create these last two things when writing new missions. However, both of these are difficult to do in an automated fashion, even with an engine like Thorium Nova. This leads to the first question:

- What can we do with viewscreens to make it easier for crews to know what is happening at that point in a mission?

Some tactical displays had a permanent overlay showing information about the ship itself - shield status, hull damage, current destination and speed, and the damaged systems. Maybe other things? - I can't remember exactly what it had.

### Thorium Nova Prime

My previous iteration of Thorium Nova did have a viewscreen, which just showed a view outside the ship. This provides a lot of context in and of itself - is the ship around a planet, or in the middle of space? Is the ship moving or stationary? Is the ship under attack? I personally think we should do everything we can to always show the exterior view on the viewscreen.

To provide the impression of movement, a warp stars effect should be used when the ship is at warp speed, and a moving space dust effect should be used at impulse speed. A skybox is also an important affordance to indicate the ship is rotating. This has already been implemented in the star map editor using a method demonstrated here: https://wwwtyro.net/2016/10/22/2D-space-scene-procgen.html

Objects should be shown at real-life sizes and distances. This includes stars, planets, moons, space stations, and ships. However, this makes it incredibly difficult to see them on the viewscreen unless the crew's ship is very close to the objects. This leads to the first question:

- How can we make it easier for crews to identify small or far-away objects on the viewscreen? Some kind of indicator or overlay perhaps?

Another thing that was displayed was Navigation Waypoints. The Navigator could choose objects and points of space to designate as waypoints which could then be used as guides to rotate the ship on course and activate the autopilot. The waypoints were included on the viewscreen to add another point of reference when rotating the ship to set course. It looked something like this:

![Viewscreen Waypoints](https://files.thoriumsim.com/file/thorium-public/blog-images/update-22-navigation/viewscreen.jpeg)

### Other ideas for the viewscreen

A few other things that would be cool to include:

- Projectiles and beam weapons for weapons fire.
- Navigational hazards, like asteroid fields, black holes, and nebulas.
- Crew controls for choosing what overlays appear on the viewscreen (showing and hiding the waypoints, as one example)
- Crew controls for changing the viewscreen camera view
- Putting the video camera feed from a probe on the viewscreen (crew controlled, of course)
- Showing helpful overlays about different things on the viewscreen, like statistics about a selected planet or ship
- Alerts or warnings - red alert status, collision alert, power grid problems, target locks, etc. - Basically conveying as much information as possible to the crew.
- Permanent overlays showing the status of the ship
- Notifications that pop up indicating that something
- Occasional overlays showing something that is currently happening. One idea is communications status - a nice graphic showing that there is an active communication, and who the communication is with.

As a reminder, I would prefer to have the viewscreen showing what's outside the ship as often as possible. That's why many things listed above are overlays, shown on top of the viewscreen. But it definitely should be possible to show other things - even arbitrary things that are defined as part of the mission. These could include pictures or videos that are bundled with the mission.

### Summary

The viewscreen is a landmark which the crew can use to know what is happening during a mission. It should prioritize showing what is going on outside the ship, but overlays can supplement that with additional relevant displays.

Questions to be answered:
- What can we do with viewscreens to make it easier for crews to know what is happening at that point in a mission?
- How can we make it easier for crews to identify small or far-away objects on the viewscreen? Some kind of indicator or overlay perhaps?
- What other things would be cool to show on the viewscreen, either as part of the exterior view, as an overlay, or as something else entirely?

I look forward to hearing your thoughts!

**Comment by astrolamb-gaming** [comment]
Some great ideas here; I don't have a lot to add, but here's my few thoughts:

1. There could be a side panel with what effectively amount to current primary objectives. To allow the crew options rather than force them towards particular objectives, these could be defined by the CO/XO. Perhaps a basic log of what's beend done so far could be included as well?
2. I like the indicators idea a lot. There could be different icons for different object types, e.g. one for planets, another for nearby ships (probably based on faction), another for waypoints or points of interest. We could also have indicators during combat for the weapons officer's targets, ships targeting the player ship, and incoming missiles.

**Reply by alexanderson1993** [reply]
Objectives are a great idea. I think how they work deserves its own discussion, but having the option to include it on the viewscreen is a great idea.

Great ideas for indicators too. Outlining some of the possible indicators makes it easier to know what to implement when the time comes.

One thing that I want to make sure is clear - the CO/XO should have a lot of power to control what is and is not on the viewscreen. Each of the layers - the individual overlays, the alerts, the indicators - should be toggleable so the crew can choose exactly what they want to see and what noise they want to hide. (though I admit giving the crew control of this does undermine the "give crew members mission context" purpose of viewscreens, but I think its a fair compromise, and gives the XO something meaningful to do)

**Comment by alexanderson1993** [comment]
It's important to point out that there was also a previous discussion about viewscreens, which I still largely agree with: https://github.com/Thorium-Sim/thorium-nova/discussions/41
